### PR TITLE
Model injection

### PR DESCRIPTION
**Pull request by aidan-curtis**

### CODE REVIEW

File: lisdf/parsing/parse_sdf.py [hunk 1]
```diff
@@ -1,23 +1,52 @@
...
+    files with absolute paths so the sdf object is self-contained.
+    """
+    for wi, world in enumerate(sdf.aggregate_order):
+        for model in world.models:
+            for component in model.aggregate_order:
+                if isinstance(component, Link):
+                    for link_component in component.aggregate_order:
+                        if isinstance(link_component, Collision):
+                            if (
+                                "geometry" in link_component.__dict__.keys()
```
Comment (williamshen-nz) [inline lisdf/parsing/parse_sdf.py:19]: A `Collision` object is guaranteed to have a `self.geometry` but that might be None. What's the point of the check here?

Same story for `Visual`

File: lisdf/parsing/parse_sdf.py [hunk 2]
```diff
@@ -1,23 +1,52 @@
...
+                                    model_path, link_component.geometry.uri
+                                )
+                        if isinstance(link_component, Visual):
+                            if (
+                                "geometry" in link_component.__dict__.keys()
+                                and isinstance(link_component.geometry, Mesh)
+                            ):
+                                link_component.geometry.uri = os.path.join(
+                                    model_path, link_component.geometry.uri
+                                )
```
Comment (williamshen-nz) [inline lisdf/parsing/parse_sdf.py:32]: It looks like the code within the `if` checks do the same thing.

Would `isinstance(link_component, (Collision, Visual))` work here instead?

File: tests/test_parsing/test_parse_sdf.py [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 
 def test_recursive_sdf_parsing():
-    parsed_sdf = load_sdf("mud_test.sdf")
+    parsed_sdf = load_sdf("mud_test/model.sdf")
     world = parsed_sdf.aggregate_order[0]
     assert len(world.includes) == 5
     assert len(world.models) == 7
```
Comment (williamshen-nz) [inline tests/test_parsing/test_parse_sdf.py:8]: Can you add some asserts checking the injected paths are correct?

File: lisdf/parsing/parse_sdf.py [hunk 4]
```diff
@@ -1,23 +1,52 @@
...
+    This function replace relative paths to object and material
+    files with absolute paths so the sdf object is self-contained.
+    """
+    for wi, world in enumerate(sdf.aggregate_order):
+        for model in world.models:
+            for component in model.aggregate_order:
+                if isinstance(component, Link):
+                    for link_component in component.aggregate_order:
+                        if isinstance(link_component, Collision):
+                            if (
```
Comment (williamshen-nz) [inline lisdf/parsing/parse_sdf.py:18]: Can you break this apart into a few more methods and collapse the if statements? There's way too much nesting here...

File: lisdf/parsing/parse_sdf.py (hunk 2 above)
Comment (williamshen-nz) [reply lisdf/parsing/parse_sdf.py:32]: Also are there any other types we would want to inject the URIs? Looking at sdf documentation there is quite a lot of different elements but they may not be relevant for us: http://sdformat.org/spec?elem=link

**Review by NishanthJKumar — APPROVED** [review]
LGTM, but I now see that Will raised some really good points - be sure to address before merging!

File: lisdf/parsing/parse_sdf.py (hunk 2 above)
Comment (NishanthJKumar) [reply lisdf/parsing/parse_sdf.py:32]: I didn't notice the copied code in the `if` - I agree with Will that it'd be preferable to merge these into a single statement!

File: tests/test_parsing/test_parse_sdf.py (hunk 3 above)
Comment (NishanthJKumar) [reply tests/test_parsing/test_parse_sdf.py:8]: +1

File: lisdf/parsing/parse_sdf.py (hunk 4 above)
Comment (NishanthJKumar) [reply lisdf/parsing/parse_sdf.py:18]: I agree that this is quite a nested function, but I feel like it's more readable/understandable in this form than if broken up - what do you think @aidan-curtis ?

File: lisdf/parsing/parse_sdf.py [hunk 5]
```diff
@@ -1,23 +1,50 @@
...
+    model_path = os.path.join(models_dir, os.path.dirname(model_name))
+    return inject_absolute_path(sdf, model_path)
 
 
 if __name__ == "__main__":  # pragma: no cover
-    sdf_test = "mud_test.sdf"
+    sdf_test = "mud_test/model.sdf"
     sdf_results = load_sdf(sdf_test)
-    print(sdf_results)
+    # print(sdf_results)
```
[the remark this answers is not in the document]
Comment (williamshen-nz) [reply lisdf/parsing/parse_sdf.py:50]: It was useful for me to load the sdf and debug with pdb

**Pull request merged**